qiskit/providers/backend.py

```diff
@@ -229,3 +241,445 @@ class can handle either situation.
+            self._options.update_config(**fields)
+        self._basis_gates_all = None
+        self.name = name
+        self.description = description
+        self.online_date = online_date
+        self.backend_version = backend_version
+
+    @property
+    def instructions(self) -> List[Tuple[Instruction, Tuple[int]]]:
+        """A list of Instruction tuples on the backend of the form ``(instruction, (qubits)``"""
+        return self.target.instructions
+
+    @property
+    def operations(self) -> List[Instruction]:
+        """A list of :class:`~qiskit.circuit.Instruction` instances that the backend supports."""
+        return list(self.target.operations)
+
+    def _compute_incomplete_basis(self):
+        incomplete_basis_gates = []
+        all_permutations_all_size = {}
+        for inst, qarg_props in self.target.items():
+            qargs = set(qarg_props)
+            size = len(next(iter(qargs)))
+            if size not in all_permutations_all_size:
+                all_permutations_all_size[size] = set(
+                    itertools.permutations(range(self.target.num_qubits), size)
+                )
+            if qargs != all_permutations_all_size[size]:
+                incomplete_basis_gates.append(inst)
+        self._basis_gates_all = incomplete_basis_gates
+
+    @property
+    def operation_names(self) -> List[str]:
+        """A list of instruction names that the backend supports."""
+        if self._basis_gates_all is None:
+            self._compute_incomplete_basis()
+        if self._basis_gates_all:
+            invalid_str = ",".join(self._basis_gates_all)
+            msg = (
+                f"This backend's operations: {invalid_str} only apply to a subset of "
+                "qubits. Using this property to get 'basis_gates' for the "
+                "transpiler may potentially create invalid output"
+            )
+            logger.warning(msg)
+        return list(self.target.operation_names)
+
+    @property
+    @abstractmethod
+    def target(self):
+        """A :class:`qiskit.transpiler.Target` object for the backend.
+
+        :rtype: Target
+        """
+        pass
+
+    @property
+    def num_qubits(self) -> int:
+        """Return the number of qubits the backend has."""
+        return self.target.num_qubits
+
+    @property
+    def coupling_map(self):
+        """Return the :class:`~qiskit.transpiler.CouplingMap` object"""
+        return self.target.coupling_map()
+
+    @property
+    def instruction_durations(self):
+        """Return the :class:`~qiskit.transpiler.InstructionDurations` object."""
+        return self.target.durations()
+
+    @property
+    @abstractmethod
+    def max_circuits(self):
+        """The maximum number of circuits (or Pulse schedules) that can be
+        run in a single job.
+
+        If there is no limit this will return None
+        """
+        pass
+
+    @classmethod
+    @abstractmethod
```

Comment:
Does this need to be abstract, or can it just return empty `Options`?

Reply:
An empty `Options()` can be returned but this method needs to be defined by a subclass because it dictates how a backend represents user tunable options and it's called by `__init__()` to initialize the default option. If they have none just doing:

```
@classmethod
def _default_options(cls):
    return Options()
```